wallet/src/lib.rs

```diff
@@ -103,6 +108,109 @@ fn set_name(name: String) {
+struct Assets {
+    contents: HashMap<&'static str, (Vec<HeaderField>, &'static [u8])>,
+    hashes: AssetHashes,
+}
+
+impl Default for Assets {
+    fn default() -> Self {
+        Self {
+            hashes: AssetHashes::default(),
+            contents: HashMap::default(),
+        }
+    }
+}
+
+type AssetHashes = RbTree<&'static str, Hash>;
+type HeaderField = (String, String);
+
+#[derive(Clone, Debug, CandidType, Deserialize)]
+struct HttpRequest {
+    method: String,
+    url: String,
+    headers: Vec<(String, String)>,
+    body: ByteBuf,
+}
+
+#[derive(Clone, Debug, CandidType, Deserialize)]
+struct HttpResponse {
+    status_code: u16,
+    headers: Vec<HeaderField>,
+    body: Cow<'static, Bytes>,
+    streaming_strategy: Option<StreamingStrategy>,
+}
+
+#[derive(Clone, Debug, CandidType, Deserialize)]
+enum StreamingStrategy {
+    Callback { callback: Func, token: Token },
+}
+
+#[derive(Clone, Debug, CandidType, Deserialize)]
+struct Token {}
+
+#[query]
+fn http_request(req: HttpRequest) -> HttpResponse {
+    let parts: Vec<&str> = req.url.split('?').collect();
+    let asset = parts[0];
+    let assets = storage::get::<Assets>();
+    let certificate_header = make_asset_certificate_header(&assets.hashes, asset);
+    match assets.contents.get(asset) {
+        Some((headers, value)) => {
+            let mut headers = headers.clone();
+            headers.push(certificate_header);
+
+            HttpResponse {
+                status_code: 200,
+                headers,
+                body: Cow::Borrowed(Bytes::new(value)),
+                streaming_strategy: None,
+            }
+        }
+        None => HttpResponse {
+            status_code: 404,
+            headers: vec![certificate_header],
+            body: Cow::Owned(ByteBuf::from(format!("Asset {} not found.", asset))),
+            streaming_strategy: None,
+        },
+    }
+}
+
+fn make_asset_certificate_header(asset_hashes: &AssetHashes, asset_name: &str) -> (String, String) {
+    let certificate = data_certificate().unwrap_or_else(|| {
+        trap("data certificate is only available in query calls");
+    });
+    let witness = asset_hashes.witness(asset_name.as_bytes());
+    let hash_tree = ic_certified_map::labeled(b"http_assets", witness);
+    let mut serializer = serde_cbor::ser::Serializer::new(vec![]);
+    serializer.self_describe().unwrap();
+    hash_tree.serialize(&mut serializer).unwrap();
+    (
+        "IC-Certificate".to_string(),
+        format!(
+            "certificate=:{}:, tree=:{}:",
+            base64::encode(&certificate),
+            base64::encode(&serializer.into_inner())
+        ),
+    )
+}
+
+fn init_assets() {
+    let assets = storage::get_mut::<Assets>();
+    for_each_asset(|name, headers, contents, hash| {
+        if name == "/index.html" {
+            assets.hashes.insert("/", *hash);
+            assets
+                .contents
+                .insert("/", (headers.clone().clone().clone(), contents));
```

Comment:
> headers.clone().clone().clone()

I don't quite get why we need to clone 3 times

Reply:
vim typeo, fixed